codepropertygraph/src/main/scala/io/shiftleft/passes/CpgPass.scala

```diff
@@ -148,74 +152,77 @@ private[passes] case class IdentityHashWrapper[T <: AnyRef](value: T) {
   * to existing serialized CPGs as graph overlays.
   * */
 private[passes] class DiffGraphProtoSerializer() {
-  import DiffGraph._
+  private val logger = LogManager.getLogger(getClass)
 
   /**
     * Generates a serialized graph overlay representing this graph
     * */
   def serialize(appliedDiffGraph: AppliedDiffGraph): CpgOverlay = {
     implicit val builder = CpgOverlay.newBuilder()
     implicit val graph = appliedDiffGraph
-    addNodes()
-    addEdges()
-    addNodeProperties()
-    addEdgeProperties()
-    builder.build()
+    val diff = appliedDiffGraph.diffGraph
+    diff.iterator.foreach {
+      case c: DiffGraph.Change.CreateEdge       => addEdge(c, builder, appliedDiffGraph)
+      case DiffGraph.Change.CreateNode(newNode) => addNode(builder, newNode, appliedDiffGraph)
+      case DiffGraph.Change.SetNodeProperty(node, key, value) =>
+        addNodeProperty(node.getId, key, value, builder, appliedDiffGraph)
+      case DiffGraph.Change.SetEdgeProperty(_, _, _) => ???
+    }
+    val overlay = builder.build()
+    overlay
   }
 
-  private def addNodes()(implicit builder: CpgOverlay.Builder, appliedDiffGraph: AppliedDiffGraph) = {
-    appliedDiffGraph.diffGraph.nodes.foreach { node =>
-      val nodeId = appliedDiffGraph.nodeToGraphId(node)
+  private def addNode(implicit builder: CpgOverlay.Builder, node: NewNode, appliedDiffGraph: AppliedDiffGraph): Unit = {
+    val nodeId = appliedDiffGraph.nodeToGraphId(node)
 
-      val nodeBuilder = CpgStruct.Node
-        .newBuilder()
-        .setKey(nodeId)
-        .setType(NodeType.valueOf(node.label))
-
-      node.properties
-        .foreach {
-          case (key, value) if !key.startsWith("_") =>
-            val property = nodeProperty(key, value)
-            nodeBuilder.addProperty(property)
-        }
+    val nodeBuilder = CpgStruct.Node
+      .newBuilder() // TODO: Can we cache builders???
+      .setKey(nodeId)
+      .setType(NodeType.valueOf(node.label))
 
-      val finalNode = nodeBuilder.build()
-      builder.addNode(finalNode)
-    }
+    node.properties
+      .foreach {
+        case (key, value) if !key.startsWith("_") =>
```

Comment:
did you test this somewhere? this will throw a MatchError for all other cases (`if key.startsWith("_")`)
demo:
```
List(1,2,3).foreach { case i if i < 3 => println(i) }
1
2
scala.MatchError: 3 (of class java.lang.Integer)
```

Reply:
This is the way it is now, I've just moved this code from few lines above.

Reply:
ok, so we expect this to crash and burn when the property name starts with `_`? sounds a bit odd.
i guess it's ok because no diffgraph would/should ever create such a property, but i'm not sure why we'd do this check in the first place. should we better just remove the `if !key.startsWith("_")` condition @fabsx00 ?